Rework volume reconstruction

jsafrane opened the pull request:
#### What type of PR is this?
Somewhere feature and cleanup - it should not change kubelet behavior (too much) and it's necessary for an upcoming feature.

/kind cleanup
/kind feature

#### What this PR does / why we need it:
Right now, Kubelet reconstructs volumes that are mounted on the host only after kubelet's desired state of world (DSW) was populated. And it reconstructed only volumes that are not in the DSW (because those volumes will be "fixed" by mounting the volumes in the usual way).

Split volume reconstruction into three distinct steps:
1. Reconstruct **all** mounted volumes on the host **right when kubelet starts**.
    * Reconstruct only information that do not need access to the API server.
    * Add these volumes to actual state of world (AWS) as *uncertain*.
    * Keep record of the volumes that failed reconstruction and all reconstructed volumes.
  
2. After DesiredStateOfWorld is fully populated and kubelet can actually check what volumes are needed, force clean up volumes that failed reconstruction and are not needed by any pods. (This is done also without this PR).

3. After DesiredStateOfWorld is fully populated, try to update `devicePaths` of reconstructed volumes from `node.status.volumesAttached`, because devicePaths obtained from reconstruction may not be accurate.
    * This requires connection to the API server, which can be established long after kubelet start.

I renamed few functions / variables on the way to better match their purpose and hidden all changed functionality in reconstruct_new.go + [behind SELinuxMountReadWriteOncePod feature gate](https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/1710-selinux-relabeling#feature-enablement-and-rollback). This is a pre-requisite of https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/1710-selinux-relabeling, where we need to know SELinux label of volume mount in ASW and compare it with desired SELinux label from desired state of world (DSW). The current reconstruction (put only volumes that need to be unmounted to ASW) will not populate the SELinux label from existing mounts.

TODO:
* Exp. backoff for retrieving Node from the API server.

#### Special notes for your reviewer:
Tested with in-tree AWS EBS, in-tree iSCSI and CSI AWS EBS.

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Kubelet now reconstructs its full cache of mounted volumes after restart; previously it reconstructed only volumes that were not used by any pod to be able to unmount them.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
[KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/1710-selinux-relabeling
```

jsafrane committed:
Rework volume reconstruction

Subsequent SELinux work (see http://kep.k8s.io/1710) will need
ActualStateOfWorld populated around the time kubelet starts mounting
volumes.

Therefore reconstruct volumes before starting reconciler, but do not depend
on the desired state of world populated nor node.status - both need a
working API server, which may not be available at that time.

All reconstructed volumes are marked as Uncertain and reconciler will sort
them out - call SetUp to ensure the volume is really mounted when a pod
needs the volume or call TearDown then there is no such pod.

Finish the reconstruction when the API server becomes available:
- Clean up volumes that failed reconstruction and are not needed.

- Update devicePath of reconstructed volumes from node.status. Make sure
  not to overwrite devicePath that may have been updated when the volume
  was mounted by reconcile().

Hiding all this rework behind SELinuxMountReadWriteOncePod FeatureGate,
just to make sure we have a way back if this commit is buggy.

jsafrane commented:
cc @jingxu97 @gnufied, trying to rework volume reconstruction as we discussed yesterday. To me it looks working, tested in local-up-cluster only and limited nr. of volumes (in-tree AWS EBS, CSI AWS EBS and in-tree iSCSI).

jsafrane commented:
Some test failures may be genuine
> [wait-control-plane] Waiting for the kubelet to boot up the control plane as static Pods from directory "/etc/kubernetes/manifests". This can take up to 4m0s
> ...
> Unfortunately, an error has occurred:
	timed out waiting for the condition

jsafrane commented:
Hmm, kind runs kube-apiserver as a static pod, i.e. there is no API server when kubelet starts. With this PR, kubelet needs to populate desired state of the world before starting any pod, i.e. it needs the API server.

gnufied commented on pkg/kubelet/volumemanager/cache/actual_state_of_world.go:
```diff
@@ -525,7 +525,10 @@ func (asw *actualStateOfWorld) AddPodToVolume(markVolumeOpts operationexecutor.M
+		// Add new mountedPod or update existing uncertain one - the new markVolumeOpts may
+		// have updated information. Especially reconstructed volumes (marked as uncertain
+		// during reconstruction) need update.
 		podObj = mountedPod{
```
Does *not* doing this results in real breakage? (sorry just curious, I suspected it will but I don't know for sure).

jsafrane replied:
I don't know, you told me that we can't trust reconstructed volumes. And I think we should not trust Uncertain volumes either, so I update everything.

gnufied commented on pkg/kubelet/volumemanager/reconciler/reconciler.go:
```diff
@@ -143,38 +143,35 @@ type reconciler struct {
-			rc.sync()
-		}
-	}
+	go rc.sync()
```
So basically reconcile can run parallel to reconstruction and we do not wait for DSOW to be populate before doing reconstruction? This means for any volume type, where reconstruction fails, the volume/mount point may leak.

jsafrane replied:
I see, reconstruction needs DSW populated here to force unmount volumes where the reconstruction itself failed: https://github.com/kubernetes/kubernetes/blob/f0d5ea1e1d4d43cd5caac92800207fcd2a62538c/pkg/kubelet/volumemanager/reconciler/reconciler.go#L377-L390

So, back to the drawing board. I was thinking about adding just an "uncertain tombstone" to ASW, where the reconstruction sync would just mark that there is a dir for the volume in /var/lib/kubelet/pods and the actual reconstruction would happen in UnmountVolume / MountVolume operations. We could try reconstruction couple of times and if it fails for too long, then force unmount. IMO it would be cleaner that today's approach (try once), but it's also more complicated.

jsafrane replied:
>  I was thinking about adding just an "uncertain tombstone" to ASW, where the reconstruction sync would just mark that there is a dir for the volume in /var/lib/kubelet/pods and the actual reconstruction would happen in UnmountVolume / MountVolume operations

Tried that, failed fast. To mark anything as uncertain, kubelet needs UniqueVolumeID and that is not available before reconstruction completes.

gnufied commented:
Did the disruptive tests we discussed are passing with this PR btw? I don' think they are running always.

msau42 commented:
/assign @jingxu97

jsafrane commented:
/hold
This PR does not work
* Reconstruction still needs DSW populated, https://github.com/kubernetes/kubernetes/pull/108180#pullrequestreview-886294681. That is hard to fix.
* Since reconstruction can happen after reconciler finishes mounting of a volume, it may overwrite a fully mounted volume with uncertain. That can be fixed relatively easily.

jsafrane commented:
I split reconstruction into two steps. Reconstruct what can be reconstructed without talking to the API server before `reconcile()` starts and add volumes that failed reconstruction to a list. Process that list only after DSW is populated and force unmount these volumes.

There is still one TODO item, after reconstruction succeeds, kubelet used to repace reconstructed `devicePath` from `node.status`. This is commented out in this PR, it needs to be filed when kubelet has connection to the API server, most probably around UnmountVolume / UnmountDevice time.

jingxu97 commented on pkg/kubelet/volumemanager/reconciler/reconciler.go:
```diff
@@ -354,67 +346,49 @@ type reconstructedVolume struct {
-	volumesNeedUpdate := make(map[v1.UniqueVolumeName]*reconstructedVolume)
-	volumeNeedReport := []v1.UniqueVolumeName{}
+	reconstructedVolumes := make(map[v1.UniqueVolumeName]*reconstructedVolume)
+	volumesInNodeStatus := []v1.UniqueVolumeName{}
```
after the change above, reconstruct is performed first, before any reconcile operations, then actual state should be empty, so no need for this check?

jingxu97 commented on pkg/kubelet/volumemanager/reconciler/reconciler.go:
```diff
@@ -602,11 +576,12 @@ func getDeviceMountPath(volume *reconstructedVolume) (string, error) {
+func (rc *reconciler) updateStates(reconstructedVolumes map[v1.UniqueVolumeName]*reconstructedVolume) error {
 	// Get the node status to retrieve volume device path information.
-	rc.updateDevicePath(volumesNeedUpdate)
+	// TODO: see what volume plugins need this and move to UnmountVolume
```
you plan to work on this in the same PR or a different one?

jsafrane replied:
I added devicePath update from nodeStatus.

jingxu97 commented on pkg/kubelet/volumemanager/reconciler/reconciler.go:
```diff
@@ -140,41 +141,42 @@ type reconciler struct {
-	rc.unmountDetachDevices()
+	// Unmount volumes only when DSW is fully populated to prevent unmounting a volume
+	// that is still needed, but it did not reach DSW yet.
+	if dswPopulated {
```
actually this block can be combined with line 157. I understand this ordering was there before, not sure the exact reason for it.

jsafrane commented:
I updated the PR description a lot, together with changing the implementation. Now the reconstruction has 3 separate steps:
1. Reconstruct whatever can be done before reconciliation starts.
2. After DSW is populated, force cleanup all volumes that failed reconstruction and are not needed any longer.
3. After connection to the API server is established, update devicePaths from node.status.

jingxu97 commented on pkg/kubelet/volumemanager/reconciler/reconciler.go:
```diff
@@ -140,41 +142,61 @@ type reconciler struct {
-			rc.sync()
-		}
-	}
+	rc.reconstructVolumes()
```
Since reconstructVolumes() is executed before reconciler even starts, we need to make sure it does not block. Checking he code there, so far seems fine. There are a few sys calls like os.Open(), I only see issue related to open named pipefile cause process hang. https://github.com/golang/go/issues/33050

In case the large number of existing mounts, it might cause slight delay on reconciler.

jsafrane replied:
In theory, any I/O can block when the storage backend is not available (like /var/lib/kubelet is on a NFS server or a network disk). In practice, if `open()` blocks, then the machine has more serious problems than reconstruction.

I went through all volume plugins and none of them seem to issue any API calls. They're parsing mount tables, json files and similar stuff. A second pair of eyes is welcome!

gnufied replied:
iirc some NFS volumes can cause permanent locks in case nfs-server was not available.

jingxu97 commented on pkg/kubelet/volumemanager/reconciler/reconciler.go:
```diff
@@ -568,25 +563,11 @@ func (rc *reconciler) reconstructVolume(volume podVolume) (*reconstructedVolume,
 		blockVolumeMapper: volumeMapper,
 	}
+	// Remember to restore devicePath
+	rc.volumesNeedDevicePath = append(rc.volumesNeedDevicePath, uniqueVolumeName)
```
why this is needed? how about linen 410?

jsafrane replied:
Good catch, deleted. I kept it at line 410 / `reconstructVolumes()`

jsafrane commented on pkg/kubelet/volumemanager/reconciler/reconciler.go:
```diff
@@ -140,41 +142,61 @@ type reconciler struct {
+	}
+
+	if !rc.devicePathsReconstructed() {
+		klog.Infof("JSAF: trying to update device paths")
```
Sorry, it's an leftover debug log. Removed.

jsafrane committed:
Add unit tests for UpdateReconstructedDevicePath

jsafrane committed:
Add unit test for reconstructVolumes()

gnufied commented:
Should we feature gate this change for SELinux change and make the new recontruction logic as default when SELinux feature goes beta?

jsafrane commented:
> Should we feature gate this change for SELinux change and make the new recontruction logic as default when SELinux feature goes beta?

We discussed that on our CSI call and I don't see an easy way how to have two different `reconstructVolumes()` and `Run()` functions switched by a feature gate. I would basically need a copy of these.

jsafrane committed:
Add SELinuxMountReadWriteOncePod feature gate

jsafrane committed:
Move all volume reconstruction code into separate files

There is no code change, just moving code around and preparing for the
subsequent commit.

jsafrane commented:
Added a feature gate and reorganized the code a bit. The old functionality is now in `reconstruct.go`, all the functionality is in `reconstruct_new.go` and feature gate SELinuxMountReadWriteOncePod switches between those two.

jsafrane commented:
With 27 volumes (AWS EBS max. attach limit for a single node), a small `hack/local-up-cluster.sh` cluster and a static pod (they get started early, even before kubelet can talk to the API server)

* With no volumes to reconstruct: 
    ```
    I0310 16:56:52.273862  135172 flags.go:64] FLAG: --add-dir-header="false"
    I0310 16:56:53.498508  135172 reconciler.go:244] "Starting operationExecutor.MountVolume for volume \"podinfo\" (UniqueName: \"kubernetes.io/downward-api/24543fa0a6a22631474a6c29d14a1ed7-podinfo\") pod \"statictestpod-ip-172-18-7-58.ec2.internal\" (UID: \"24543fa0a6a22631474a6c29d14a1ed7\") " pod="default/statictestpod-ip-172-18-7-58.ec2.internal"
    ```

* With 27 EBS volumes to reconstruct + SELinuxMountReadWriteOncePod enabled:
    ```
    I0310 17:10:53.485295  150216 flags.go:64] FLAG: --add-dir-header="false"
    I0310 17:10:54.905184  150216 reconciler.go:244] "Starting operationExecutor.MountVolume for volume \"podinfo\" (UniqueName: \"kubernetes.io/downward-api/24543fa0a6a22631474a6c29d14a1ed7-podinfo\") pod \"statictestpod-ip-172-18-7-58.ec2.internal\" (UID: \"24543fa0a6a22631474a6c29d14a1ed7\") " pod="default/statictestpod-ip-172-18-7-58.ec2.internal"
    ```

* With 27 EBS volumes to reconstruct + SELinuxMountReadWriteOncePod disabled:
    ```
    I0310 18:04:14.458200  202068 flags.go:64] FLAG: --add-dir-header="false"
    I0310 18:04:16.258946  202068 reconciler.go:244] "Starting operationExecutor.MountVolume for volume \"podinfo\" (UniqueName: \"kubernetes.io/downward-api/24543fa0a6a22631474a6c29d14a1ed7-podinfo\") pod \"statictestpod-ip-172-18-7-58.ec2.internal\" (UID: \"24543fa0a6a22631474a6c29d14a1ed7\") " pod="default/statictestpod-ip-172-18-7-58.ec2.internal"
    ```

I did just one attempt for each case, the numbers may be quite random. It's weird that the old reconciler is the slowest.

jsafrane commented:
/retest
curious about storage-slow tests

gnufied commented on pkg/kubelet/volumemanager/reconciler/reconstruct_new.go:
```diff
@@ -0,0 +1,224 @@
+		}
+		// The reconstructed volumes are mounted, hence a previous kubelet must have already put it into node.status.volumesInUse.
+		// Remember to update DSW with this information.
+		rc.desiredStateOfWorld.MarkVolumesReportedInUse(reconstructedVolumeNames)
```
we are calling `MarkVolumesReportedInUse` without potentially populating DSOW. Does this have any effect? It seems like the function won't do anything and will be a NO-OP.

jsafrane replied:
Good catch! Added a new `rc.volumesNeedReportedInUse` with volumes to mark as in use *after* DSW is populated.

gnufied commented on pkg/kubelet/volumemanager/reconciler/reconstruct_new.go:
```diff
@@ -0,0 +1,224 @@
+	klog.V(2).InfoS("Volume reconstruction finished")
+}
+
+func (rc *reconciler) updateStatesNew(reconstructedVolumes map[v1.UniqueVolumeName]*reconstructedVolume) error {
```
nit: we never return an error from this function, did we intend to?

jsafrane committed:
Remove devicePathsReconstructed

jsafrane committed:
Remove error return value from updateStatesNew

It always return nil anyway.

jsafrane commented on pkg/kubelet/volumemanager/reconciler/reconstruct_new.go:
```diff
@@ -0,0 +1,224 @@
+		}
+		klog.V(4).InfoS("Volume is marked as uncertain and added into the actual state", "pod", klog.KObj(volume.pod), "podName", volume.podName, "volumeName", volume.volumeName)
+		// If the volume has device to mount, we mark its device as mounted.
+		if volume.deviceMounter != nil || volume.blockVolumeMapper != nil {
```
It's the same as the original `updateStates()`. IMO the order does not matter at all, both will be called, because all errors are ignored.

jsafrane commented on pkg/kubelet/volumemanager/reconciler/reconstruct_new.go:
```diff
@@ -0,0 +1,224 @@
+		rc.cleanOrphanVolumes()
+	}
+
+	if !rc.devicePathsReconstructed() {
```
`devicePathsReconstructed` removed.

jsafrane committed:
Mark volumes as in use only after DSW is populated

To be able to mark volumes in DSW as marked "in use" in node.status, the
whole DSW needs to be populated first.

jsafrane commented on pkg/kubelet/volumemanager/cache/actual_state_of_world.go:
```diff
@@ -525,7 +549,16 @@ func (asw *actualStateOfWorld) AddPodToVolume(markVolumeOpts operationexecutor.M
+		// an update.
+		updateUncertainVolume = utilfeature.DefaultFeatureGate.Enabled(features.SELinuxMountReadWriteOncePod) && podObj.volumeMountStateForPod == operationexecutor.VolumeMountUncertain
+	}
+	if !podExists || updateUncertainVolume {
```
I'm not sure I follow, can you please elaborate?

gnufied replied:
So there was this bug - https://github.com/kubernetes/kubernetes/issues/103143 which was reported because during reconstruction we do not get correct `outerSpecName` and if populator loop did not add volume to DSOW because of api-server unavailability then we never properly mount the volume.

After your change - it looks like the entire data structure stored in ASOW is restored for uncertain volumes, so we still need `SyncReconstructedVolume` ?

jsafrane replied:
You're correct, with `SELinuxMountReadWriteOncePod` feature enabled it should not be necessary to call `SyncReconstructedVolume`. I added a new commit that removes it.

jsafrane commented:
I tested reboot with OpenShift 4.10 (=Kubernetes 1.23.3) and with in-tree AWS EBS volumes.
Both with and without this PR (and the feature enabled), the volume reconstruction failed after reboot - EBS volume plugin needs the volume mounted to be able to find the global mount + find AWS volume ID from it.

If I left Pods in the API server during the reboot, the new kubelet started all of them.
If I removed Pods from the API server during reboot, the pod local directory got removed in both cases by our fallback cleaner and the global dir was not removed (because reconstruction failed).

This PR did not have any visible effect.

jsafrane commented:
/retest

jsafrane committed:
Remove SyncReconstructedVolume call

With the new reconstruction, AWS.MarkVolumeAsMounted will update outer spec
name with the correct value from Pod.

ehashman commented:
/milestone v1.25

cici37 commented:
Hello 👋, 1.25 Release Lead here.

Unfortunately, this enhancement did not meet the [code freeze](https://github.com/kubernetes/sig-release/blob/master/releases/release_phases.md#code-freeze) criteria because there are still unmerged k/k code PRs.

If you still wish to progress this enhancement in v1.25, please file an [exception request](https://github.com/kubernetes/sig-release/blob/master/releases/EXCEPTIONS.md). Thank you so much!

/milestone clear
